File: backendj/src/com/runsim/backend/SCTPClient.java
package com.runsim.backend;

import com.sun.nio.sctp.*;

import java.io.PrintStream;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;

public class SCTPClient {

    public void send(String hostname, int port, int protocolId, int streamNumber, byte[] data) throws Exception {
        InetSocketAddress serverAddress = new InetSocketAddress(hostname, port);
        SctpChannel channel = SctpChannel.open(serverAddress, 0, 0);
        AssociationHandler associationHandler = new AssociationHandler();

        ByteBuffer outgoingBuffer = ByteBuffer.wrap(data);
        ByteBuffer incomingBuffer = ByteBuffer.allocate(1073741824);

        MessageInfo outgoingMessage = MessageInfo.createOutgoing(null, streamNumber);
        outgoingMessage.payloadProtocolID(protocolId);

        channel.send(outgoingBuffer, outgoingMessage);

        MessageInfo messageInfo;
        while (true) {
            messageInfo = channel.receive(incomingBuffer, System.out, associationHandler);
            if (messageInfo == null || messageInfo.bytes() == -1) {
                break;
            }

            byte[] receivedBytes = new byte[messageInfo.bytes()];
            for (int i = 0; i < receivedBytes.length; i++) {
                receivedBytes[i] = incomingBuffer.get(i);
            }

            handleMessage(channel, receivedBytes);
        }

        channel.close();
    }

    private void handleMessage(SctpChannel channel, byte[] data) throws Exception {

    }

    private static class AssociationHandler extends AbstractNotificationHandler<PrintStream> {

        @Override
        public HandlerResult handleNotification(AssociationChangeNotification notification, PrintStream attachment) {
            if (notification.event() == AssociationChangeNotification.AssocChangeEvent.COMM_UP) {
                int outbound = notification.association().maxOutboundStreams();
                int inbound = notification.association().maxInboundStreams();
                attachment.printf(
                        "New association setup with %d outbound streams" + ", and %d inbound streams.\n",
                        outbound, inbound
                );
            }
            return HandlerResult.CONTINUE;
        }

        @Override
        public HandlerResult handleNotification(ShutdownNotification notification, PrintStream attachment) {
            attachment.print("The association has been shutdown.\n");
            return HandlerResult.RETURN;
        }
    }
}

File: backend/src/main/kotlin/SCTPClient.kt
import com.sun.nio.sctp.*

import java.io.*
import java.net.*
import java.nio.*

object SCTPClient {

    public fun send(hostname: String, port: Int, protocolId: Int, streamNumber: Int, data: ByteArray) {
        val serverAddress = InetSocketAddress(hostname, port)
        val channel = SctpChannel.open(serverAddress, 0, 0)
        val associationHandler = AssociationHandler()

        val outgoingBuffer = ByteBuffer.wrap(data)
        val incomingBuffer = ByteBuffer.allocate(1073741824)

        val outgoingMessage = MessageInfo.createOutgoing(null, streamNumber)
        outgoingMessage.payloadProtocolID(protocolId)

        channel.send(outgoingBuffer, outgoingMessage)

        var messageInfo: MessageInfo?
        do {
            messageInfo = channel.receive(incomingBuffer, System.out, associationHandler)
            if (messageInfo == null || messageInfo.bytes() == -1) {
                break
            }

            val receivedBytes = ByteArray(messageInfo.bytes())
            for (i in 0 until messageInfo.bytes()) {
                receivedBytes[i] = outgoingBuffer[i]
            }

            handleMessage(channel, receivedBytes)
        } while (messageInfo != null)
    }

    private fun handleMessage(channel: SctpChannel, data: ByteArray) {
        val msg = MessageInfo.createOutgoing(null, 0)
        msg.payloadProtocolID(60)
        channel.send(ByteBuffer.wrap(byteArrayOf(1, 2, 3, 4)), msg)
    }

    private class AssociationHandler : AbstractNotificationHandler<PrintStream>() {
        override fun handleNotification(not: AssociationChangeNotification, stream: PrintStream?): HandlerResult {
            if (not.event() == AssociationChangeNotification.AssocChangeEvent.COMM_UP) {
                val outbound = not.association().maxOutboundStreams()
                val inbound = not.association().maxInboundStreams()
                stream!!.printf(
                    "New association setup with %d outbound streams" + ", and %d inbound streams.\n",
                    outbound, inbound
                )
            }
            return HandlerResult.CONTINUE
        }
        override fun handleNotification(not: ShutdownNotification?, stream: PrintStream): HandlerResult {
            stream.printf("The association has been shutdown.\n")
            return HandlerResult.RETURN
        }
    }
}